### PR TITLE
fix (gcm): invalid character 'J' looking for beginning of value

### PR DESCRIPTION
**Pull request by khevse**

### CODE REVIEW

File: pkg/provider/helper.go [hunk 1]
```diff
@@ -46,3 +49,81 @@ func SendWithRetry(maxRetries int, send func() (statusCode int, _ error)) error
...
+	if err != nil {
+		return nil, err
+	}
+
+	return RemoveSecretsFromJSON(out), nil
+}
+
+var _SecretBegin = []byte(`:"`)
+
+func RemoveSecretsFromJSON(in []byte) []byte {
```
Comment (alexboyko) [inline pkg/provider/helper.go]: is it much better than replace with regex?

File: pkg/provider/helper.go [hunk 2]
```diff
@@ -46,3 +49,81 @@ func SendWithRetry(maxRetries int, send func() (statusCode int, _ error)) error
 
 	return nil
 }
+
+// DecodeJSONResponse unmarshal response in json format to the object.
+// If server returns invalid json data, the method represents a response body
+// as an error
+func DecodeJSONResponse(r io.Reader, retval interface{}) error {
```
Comment (alexboyko) [inline pkg/provider/helper.go]: we can detect if it valid first using decoder.Valid

File: pkg/provider/helper.go [hunk 3]
```diff
@@ -46,3 +49,81 @@ func SendWithRetry(maxRetries int, send func() (statusCode int, _ error)) error
...
+	decoder := json.NewDecoder(r)
+
+	err := decoder.Decode(retval)
+	if err == nil {
+		return nil
+	}
+
+	if _, ok := err.(*json.SyntaxError); ok {
+		errInfo := bytes.NewBuffer(nil)
+		if _, errCopy := io.Copy(errInfo, decoder.Buffered()); errCopy != nil {
```
Comment (alexboyko) [inline pkg/provider/helper.go]: You can use CopyN to copy no more than 2000 bytes

Comment (alexboyko) [reply pkg/provider/helper.go]: Buffered will return remaining buffer, is it expected here?

File: pkg/provider/helper.go [hunk 4]
```diff
@@ -46,3 +49,81 @@ func SendWithRetry(maxRetries int, send func() (statusCode int, _ error)) error
...
+			errInfo.Truncate(2000)
+		}
+
+		return errors.New(errInfo.String())
+	}
+
+	return err
+}
+
+func JSONWithoutSecrets(obj interface{}) ([]byte, error) {
```
Comment (alexboyko) [inline pkg/provider/helper.go]: Actually it just replaces all string values, scalars and arrays remain untouched.
Can we replace it with JSONWithValuesReplaced or similar?

File: pkg/provider/pipe.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+			return chWait
+		}():
+		}
+
+	}()
+
+	return p
+}
+
+func (p *Pipe) Read(out []byte) (n int, err error) {
```
Comment (alexboyko) [inline pkg/provider/pipe.go]: Does the io.TeeReader solve the same issue?

File: pkg/provider/helper.go (hunk 3 above)
Comment (khevse) [reply pkg/provider/helper.go]: I can't use CopyN for a short buffer. Example: https://play.golang.org/p/iAZGncjYDkX

File: pkg/provider/pipe.go (hunk 5 above)
Comment (khevse) [reply pkg/provider/pipe.go]: The pipe create another function. It is writing if anybody reading.

File: pkg/provider/helper.go (hunk 3 above)
Comment (khevse) [reply pkg/provider/helper.go]:
```
       if _, ok := err.(*json.SyntaxError); ok {
		errInfo := bytes.NewBuffer(nil)
		_, errCopy := io.CopyN(errInfo, decoder.Buffered(), 2000)
		if errCopy != nil && errCopy != io.EOF {
			return err
		}

		return errors.New(errInfo.String())
	}
```

**Pull request merged**